enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,302 @@
+internal or external connectivity. Since internal connectivity is mandatory,
+the only option is to use names for external connectivity.
+
+### Merging certificates
+Since having the node IP coexist with trust domain segregation requires not
+including it for external access, there are not too many options left for
+fulfilling the requirements.
+
+Like K3s, which is a competing product, the approach would be to merge all API
+server certificates into a single one, including service names, service IP,
+node IP, node names, and additional SANs from the configuration. The
+certificate matching problem vanishes, as well as the hostname validations and
+different CAs, as it would be a single one.
+
+This approach removes trust domains when accessing API server, like OCP does,
+but the use cases require IP connectivity and authentication from anywhere.
+
+
+### Workflow Description
+Workflow is transparent for the user, as certificates generate themselves upon
+restarting MicroShift.
+
+#### Variation [optional]
+N/A
+
+### API Extensions
+N/A
+
+### Implementation Details/Notes/Constraints [optional]
+N/A
+
+### Risks and Mitigations
+Removal of trust domains and different approach than the one in OpenShift.
+
+Having a single certificate to connect to API server means anyone can
+connect to it, and not having oauth requires extra care when configuring
+cluster roles. This, however, is not too different than the current approach.
+
+Not implementing this change makes MicroShift unable to meet requirements.
+
+### Drawbacks
+N/A
+
+## Design Details
+N/A
+
+### Open Questions [optional]
+N/A
+
+### Test Plan
+N/A
+
+### Graduation Criteria
+N/A
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+N/A
+
+### Version Skew Strategy
+N/A
+
+### Operational Aspects of API Extensions
+N/A
+
+#### Failure Modes
+N/A
+
+#### Support Procedures
+N/A
+
+## Implementation History
+N/A
+
+## Alternatives
+* Have several IPs in the node.
+   Having multiple IPs in the node would allow authentication to distinguish
+   between the destination IP and serve the adequate certificate based on what
+   was described before. This would require changing network architecture.
+* Have a single CA for all API server certificates.
+   Signing all certificates with the same CA is not different than having a
+   single certiticate. It would still need the node IP in the certificates,
+   and we have already seen that this can only be done in the internal one,
+   meaning it is also needed to add hostnames, rendering a single certificate.
+* Use localhost for internal connectivity. 
+   Using localhost as the node IP would allow internal pods use a different IP
+   than what the external connectivity needs. This would conflict with the
+   localhost certificate, which would require merging. It also invalidates
+   having more than one node.
```

Comment:
So in this approach we would still have to merge certificates, but they would be different certificates? Or a subset of the ones that this proposal talks about merging?

Reply:
How does the risk of doing that compare to merging all of them together?